pkg/scheduler/algorithm/predicates/metadata.go

```diff
@@ -527,50 +520,40 @@ func NodeLabelsMatchSpreadConstraints(nodeLabels map[string]string, constraints
 	return true
 }
 
-// returns a pointer to a new topologyPairsMaps
-func newTopologyPairsMaps() *topologyPairsMaps {
-	return &topologyPairsMaps{
-		topologyPairToPods: make(map[topologyPair]podSet),
-		podToTopologyPairs: make(map[string]topologyPairSet),
+func (m topologyPairToPods) addTopologyPair(pair topologyPair, pod *v1.Pod) {
+	if m[pair] == nil {
+		m[pair] = make(map[*v1.Pod]struct{})
 	}
+	m[pair][pod] = struct{}{}
 }
 
-func (m *topologyPairsMaps) addTopologyPair(pair topologyPair, pod *v1.Pod) {
-	podFullName := schedutil.GetPodFullName(pod)
-	if m.topologyPairToPods[pair] == nil {
-		m.topologyPairToPods[pair] = make(map[*v1.Pod]struct{})
-	}
-	m.topologyPairToPods[pair][pod] = struct{}{}
-	if m.podToTopologyPairs[podFullName] == nil {
-		m.podToTopologyPairs[podFullName] = make(map[topologyPair]struct{})
-	}
-	m.podToTopologyPairs[podFullName][pair] = struct{}{}
-}
-
-func (m *topologyPairsMaps) removePod(deletedPod *v1.Pod) {
-	deletedPodFullName := schedutil.GetPodFullName(deletedPod)
-	for pair := range m.podToTopologyPairs[deletedPodFullName] {
-		delete(m.topologyPairToPods[pair], deletedPod)
-		if len(m.topologyPairToPods[pair]) == 0 {
-			delete(m.topologyPairToPods, pair)
+func (m topologyPairToPods) removePod(deletedPod *v1.Pod, node *v1.Node) {
+	for topologyKey, topologyValue := range node.Labels {
+		pair := topologyPair{key: topologyKey, value: topologyValue}
+		if _, exist := m[pair]; exist {
```

Comment:
It might be worth caching the topology keys used in each node. But better to explore that in it's own PR.